From 44d621188e21657345e6d7cd023e5a5eaf1bb923 Mon Sep 17 00:00:00 2001
From: Sumit Singh <sumsingh@nvidia.com>
Date: Mon, 21 Apr 2014 18:12:14 +0530
Subject: [PATCH 23/27] seqlock: enhance power efficiency

Trying to improve the power efficiency in linux/seqlock.h,
using macros cpu_relaxed_read and cpu_read_relax.

Bug 1440421

Change-Id: I8dcc9dc9c72e5a6848c7823768c76f5d38e9f0d1
Signed-off-by: Sumit Singh <sumsingh@nvidia.com>
Reviewed-on: http://git-master/r/398943
(cherry picked from commit ed77ee4d50641b1a5668545c946cda681c06fcd2)
Reviewed-on: http://git-master/r/422258
GVS: Gerrit_Virtual_Submit
Reviewed-by: Sri Krishna Chowdary <schowdary@nvidia.com>
Reviewed-by: Alexander Van Brunt <avanbrunt@nvidia.com>
Signed-off-by: Jesse Chan <jc@linux.com>
Signed-off-by: Michele Beccalossi <beccalossi.michele@gmail.com>
---
 include/linux/seqlock.h | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/include/linux/seqlock.h b/include/linux/seqlock.h
index f5df8f687b4..c180829de2c 100644
--- a/include/linux/seqlock.h
+++ b/include/linux/seqlock.h
@@ -36,6 +36,7 @@
 #include <linux/preempt.h>
 #include <linux/lockdep.h>
 #include <asm/processor.h>
+#include <asm/relaxed.h>
 
 /*
  * Version using sequence counter only.
@@ -108,9 +109,9 @@ static inline unsigned __read_seqcount_begin(const seqcount_t *s)
 	unsigned ret;
 
 repeat:
-	ret = ACCESS_ONCE(s->sequence);
+	ret = cpu_relaxed_read((volatile u32 *)&s->sequence);
 	if (unlikely(ret & 1)) {
-		cpu_relax();
+		cpu_read_relax();
 		goto repeat;
 	}
 	return ret;
-- 
2.11.0

